GPS location highlighting

dg0yt opened the pull request:
![gps](https://user-images.githubusercontent.com/13567791/46854287-a4806580-ce00-11e8-8afe-afa7662938b4.gif)

lpechacek commented on src/sensors/gps_display.cpp:
```diff
@@ -178,9 +178,14 @@ void GPSDisplay::paint(QPainter* painter)
+	// Highlight markers by underlaying white under the tracking color
+	highlighting += 64;
+	if (highlighting > 2 * 192)
+		highlighting = 0;
```
Use of modulo might be more elegant at this spot.
`highlight %= 2*192+1;`

lpechacek commented on src/sensors/gps_display.cpp:
```diff
@@ -178,9 +178,14 @@ void GPSDisplay::paint(QPainter* painter)
+	highlighting += 64;
+	if (highlighting > 2 * 192)
+		highlighting = 0;
+	auto underlay = QColor{255, 255, 255, 63 + (qAbs( 192 - highlighting))};
```
The four-`int` `QColor` construction is somewhat difficult to read. How about `QColor(Qt::white).setAlpha(63 + qAbs(192 - highlighting))`?

I also don't see readily why `highlight` increases by 64 and what is the magic 192. IMHO there is too much complicated code around simple periodic alpha change. Good illustration is that I overlooked the addition of 63 in alpha value above. The parentheses around `qAbs()` are extra.

dg0yt commented:
Thanks for your review and the points you make. I will try to reorganise it in a better way. 
- For efficiency, I don't really want to decouple the highlighting variable much from of the domain of alpha value which is its final purpose. Maybe it is enough to use some expressive interface/encapsulation instead of plain int.
- With regard to power consumption, I want to avoid adding an extra timer/wake-up. We [request](http://doc.qt.io/qt-5/qgeopositioninfosource.html#updateInterval-prop) the position source to deliver updates at 1 Hz. It might be slower, but it should be close in practice. What concerns me more is if updates might be suspended when the position does not change.

jmacura commented:
From a users point of view, I have barely noticed the change of halo in the initial GIF. It might be because of the GIF quality, but I'm not sure, if this is sufficient highlight for the requester @mlerjen ?

mlerjen commented:
@jmacura: I think that will fits my request pretty well. It is mostly on darker and colorful backgrounds the issue appeared.

dg0yt commented:
> I have barely noticed the change of halo in the initial GIF. It might be because of the GIF quality, but I'm not sure

The screen resolution is <100 dpi, and the GIF has dithering. It should be more visible on mobile devices >100 dpi.

dg0yt commented:
Blinking or not, it remains quite difficult to see at 1x zoom. Is it enough to increase line width, or do we need another type of marker (pin or reverse drop)?

dg0yt commented:
Dropped the slowly pulsating white background for:
- solid white framing (but some opacity for the whole GPS display),
- bolder lines and dots (but might be too much for the signal accuracy circle)
- crosshairs (fixed size, helps a lot).
And of course it still blinks on move-to-my-location.

![crosshais](https://user-images.githubusercontent.com/13567791/46922568-2cd25680-d00b-11e8-8fad-f48d0c90b8e0.png)

dg0yt commented:
v0.8.2 vs. new implementation:  
![gps](https://user-images.githubusercontent.com/13567791/46932459-3336f180-d050-11e8-9468-9b5a634e017e.png) ![crosshais](https://user-images.githubusercontent.com/13567791/46922568-2cd25680-d00b-11e8-8fad-f48d0c90b8e0.png)

I'm quite satisfied, but of course it takes a few more pixel from the screen.

dg0yt commented:
Blinking on move-to-my-location:
![out](https://user-images.githubusercontent.com/13567791/46934364-c58ec380-d057-11e8-9b39-5c5dc5a6b9c8.gif)

dg0yt commented on src/sensors/gps_display.cpp:
```diff
@@ -291,11 +298,11 @@ void GPSDisplay::debugPositionUpdate()
+	const auto altitude = 400 + 10 * qSin(1 + offset / 10);
 	
-	MapCoordF coord(30 * qSin(0.5f * offset), 30 * qCos(0.53f * offset));
+	MapCoordF coord(30 * qSin(0.5 * offset), 30 * qCos(0.53 * offset));
```
Then it would probably turn this into `const auto coord = MapCoordF{...};`. This would make reviews even more difficult. And this particular line is debugging code which will need a refactoring anyway.

dg0yt replied:
ES.23 only asks to "Prefer the {} initializer syntax", and it points out that you must be careful with initializer lists. But it doesn't say you should prefer 
~~~
 const auto now { QTime::currentTime() };
~~~
over
~~~
 const auto now = QTime::currentTime();
~~~
And only the latter seems to allow maximum consistency, aka almost-always-auto (AAA) style, as in:
~~~
        const auto now = QTime::currentTime();
        const auto offset = now.msecsSinceStartOfDay() / qreal(10 * 1000);
        const auto accuracy = float(12 + 7 * qSin(2 + offset));
        const auto altitude = 400 + 10 * qSin(1 + offset / 10);

        const auto coord = MapCoordF{ 30 * qSin(0.5 * offset), 30 * qCos(0.53 * offset) };
~~~
(I left C/operator-style conversions for `qreal` and `float`. And I didn't go so far to use "east-side const".)
This recent article and comments are related: https://www.fluentcpp.com/2018/09/28/auto-stick-changing-style/

By the way, Qt Creator's formatting stumples over
~~~
const auto foreground = QColor{tracking_lost ? Qt::gray : Qt::red};
~~~
i.e. ternary operator in `{}? initializer. That's why I went back to `()` in this spot.

dg0yt commented on src/sensors/gps_display.cpp:
```diff
@@ -177,18 +187,19 @@ void GPSDisplay::paint(QPainter* painter)
 		painter->rotate(heading_rotation_deg);
 		
 		// Draw arrow
 		static const QPointF arrow_points[4] = {
```
Agree.

dg0yt commented on src/sensors/gps_display.h:
```diff
@@ -43,6 +42,8 @@ class MapWidget;
 /**
  * Retrieves the GPS position and displays a marker at this position on a MapWidget.
+ * 
+ * \todo Use qreal instead of bool (affects all sensor code)
```
Oops. `Use qreal instead of *float* (affects all sensor code)`

We get `qreal`  from Qt, we pass `qreal` to Qt, but internally we use `float` and do a lot of conversation for no (or little) benefit. This is a task for `dev`. I even created a Github project and card on this occassion.

lpechacek replied:
Ack.

dg0yt commented on src/sensors/gps_display.h:
```diff
@@ -119,6 +128,16 @@ private slots:
+	{
+	public:
+		const PulsatingOpacity& operator++();
+		operator qreal() const;
```
Alternatively, this might be `qreal current() const`

dg0yt commented:
> Renaming blinkTimeout() to blinkTimerHandler() and putting a brief remark in setBlinking() stating that first invocation of the timer handler starts the blinking cycle would help readability.

I will probably switch from the convienent `QTimer`/signal/slot to the lightweight `QObject::startTimer()`/`timerEvent()`.

dg0yt commented:
Major changes since previous review comments:
- in "GPSDisplay: Review, minor corrections":  
  Moved member initializations from constructor to declaration. Modified documentation comments. Removed `inline` keyword.
- in "GPSDisplay: Add blinking on move-to-my-location":  
  Switched to `QObject::timerEvent`, and elaborated `PulsatingOpacity` to a `QBasicTimer`-like interface. Achieved better separation of concerns with `PulsatingOpacity` handling a *single* cycle (one second) and `GPSDisplay` handling the number of seconds. Added documentation comments.

mlerjen commented:
Much better!
![2018-10-24-10-50-51](https://user-images.githubusercontent.com/15611653/47458042-ee8e2180-d7d8-11e8-98bf-02873ea0cbe5.png)
